### PR TITLE
refactor: add CLI config schema to make sure we don't miss options

### LINKED ISSUE #3962: command line flags don't override config file
### Describe the bug

As a tool user, I usually have a set of expectations respect to the precedence order of certain ways to set configuration values for the tools I use. One of these typical rules is that the order of application is similar to:

1. config files
2. environment variables
3. command line options/flags

where the later overrides the former (in case it is present).

In the case of Vitest, I'm finding that this is not the case when it comes to coverage related settings. For example, when I have the following config file:
```typescript
import { defineConfig } from 'vitest/config'

export default defineConfig({
	test: {
		coverage: {
			include: ['src/*.{ts,tsx}'],
			exclude: ['src/**/*.test.{ts,tsx}', 'src/stories/**/*'],
			thresholdAutoUpdate: true,
			branches: 50,
			lines: 50,
			functions: 50,
			statements: 50,
		},
	},
})
```

but I call `vitest` with `vitest run --coverage.enabled=true --coverage.thresholdAutoUpdate=false`, the later flag is not being applied, giving more importance to the one inside the config file (and as a consequence, if my coverage is higher, the config file is auto-updated to reflect the new coverage values).

### Reproduction

The previous description is enough. We just need a test with high enough coverage, and set the thresholds in our config to be below that if we want to see how the command line flag is not taking precedence over the config file value.

### System Info

```shell
System:
    OS: macOS 13.5
    CPU: (24) arm64 Apple M2 Ultra
    Memory: 14.52 GB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.1.0 - ~/Library/Caches/fnm_multishells/14573_1692095309449/bin/node
    Yarn: 3.6.1 - ~/Library/Caches/fnm_multishells/14573_1692095309449/bin/yarn
    npm: 9.6.4 - ~/Library/Caches/fnm_multishells/14573_1692095309449/bin/npm
    pnpm: 8.6.12 - ~/Library/Caches/fnm_multishells/14573_1692095309449/bin/pnpm
  Browsers:
    Chrome: 115.0.5790.170
    Edge: 114.0.1823.82
    Safari: 16.6
  npmPackages:
    @vitest/coverage-v8: ^0.34.1 => 0.34.1 
    vitest: ^0.34.1 => 0.34.1
```

### Used Package Manager

pnpm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by sheremet-va**
### Description

This PR makes it easier for us to make sure we didn't forget to add a CLI option because it requires an explicit opt-out.

This is also a continuation of #3983 to make sure all types are correct.

Closes #3983
Fixes #3962

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [ ] Ideally, include a test that fails without this PR but passes with it.
- [ ] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: packages/vitest/src/node/config/schema.ts
Comment (AriPerkkio) [inline packages/vitest/src/node/config/schema.ts]: For now I think it's ok to not (publicly) support `coverage.watermarks` via CLI. It's not commonly used command and all cases I've seen have used configuration files.

**Review by AriPerkkio — COMMENTED** [review]
Overall looks really nice. 🙌 
I think this is a solution that we can maintain quite easily. I'm glad to see the `--someOption --someOption.nestedOption=value` working without specifying the `.enabled` explicitly. 

Some quick comments below for now. I'll do proper review later this week and do some manual testing.

File: packages/vitest/src/node/cli/cac.ts
Comment (AriPerkkio) [inline packages/vitest/src/node/cli/cac.ts:215]: Do we still need these as there's `CLIOption.array` flag? Would it automatically convert these?

File: packages/vitest/src/node/cli/cac.ts
Comment (AriPerkkio) [inline packages/vitest/src/node/cli/cac.ts:196]: Should we have `CLIOption.normalize` option that would automatically do this?

File: packages/vitest/src/node/cli/cac.ts
Comment (AriPerkkio) [reply packages/vitest/src/node/cli/cac.ts:215]: Are these `enabled` setters still needed too?

https://github.com/vitest-dev/vitest/blob/dcf0d6afe974487a53f7e7c90c0f79d29fa885d2/packages/vitest/src/node/cli-api.ts#L40-L62

File: packages/vitest/src/node/cli/cac.ts
Comment (AriPerkkio) [inline packages/vitest/src/node/cli/cac.ts:103]: Are these two documented anywhere? Maybe at the end of `--help` it should print some instructions for this. 

Right now I had to read the source code before I understood why `vitest --coverage --help` didn't work.

File: packages/vitest/src/node/cli/cli-config.ts
Comment (AriPerkkio) [inline packages/vitest/src/node/cli/cli-config.ts:231]:
```suggestion
            description: 'Shortcut to set all coverage thresholds to 100 (default: false)',
```

File: packages/vitest/src/node/cli/cli-config.ts
Comment (AriPerkkio) [inline packages/vitest/src/node/cli/cli-config.ts:241]:
```suggestion
        description: 'Concurrency limit used when processing the coverage results. (default min between 20 and the number of CPUs)',
```

**Review by AriPerkkio — COMMENTED** [review]
> to make sure we didn't forget to add a CLI option because it requires an explicit opt-out.

This seems to work for top-level configurations, but not for sub-commands.

I have not yet finished reviewing this completely, but I'm confident enough about the changes. Feel free to merge this if needed. I'll look into this more later this week once I find some time.

**Comment by sheremet-va** [comment]
> This seems to work for top-level configurations, but not for sub-commands.

Can you give an example when this doesn't work? If I remove any nested commands, I get a type error.

File: packages/vitest/src/node/cli/cac.ts
Comment (sheremet-va) [reply packages/vitest/src/node/cli/cac.ts:103]: I actually have no idea what this code does 😄 , it's migrated from #3983

**Comment by AriPerkkio** [comment]
> Can you give an example when this doesn't work? If I remove any nested commands, I get a type error.

I tested this by adding new option to provider specific coverage options. But when I add an option to all providers, then it raises an error on CLI definitions. So it's got something to do with conditional types. Detecting all those combinations might be difficult so the current implementation is enough I think.

File: packages/vitest/src/node/cli/cac.ts
Comment (AriPerkkio) [reply packages/vitest/src/node/cli/cac.ts:103]: Oh right it's actually something I've written: https://github.com/vitest-dev/vitest/pull/3983#issuecomment-1694595988

I'll check how we could add instructions for this into the `--help` message.

**Comment by sheremet-va** [comment]
> I tested this by adding new option to provider specific coverage options.

Ah, yes. This only happens with `provider` because it has a complex type, and TypeScript uses the smallest common denominator when checking this.

File: packages/vitest/src/node/cli/cac.ts
Comment (sheremet-va) [reply packages/vitest/src/node/cli/cac.ts:103]: Ok, I will wait for your feedback regarding this before merging

**Review by AriPerkkio — APPROVED** [review]
Looks good to me, thanks! Also thanks to @LorenzoBloedow for earlier work on #3983! 🙌

**Comment by LorenzoBloedow** [comment]
> Also thanks to @LorenzoBloedow for earlier work on #3983! 🙌

No problem, glad you guys found a solution! If you need any help feel free to hit me up on Twitter!

File: packages/vitest/src/node/cli/cli-config.ts
Comment (hi-ogawa) [inline packages/vitest/src/node/cli/cli-config.ts:143]: I was testing multiple reporter options like this documentation https://vitest.dev/guide/reporters.html#combining-reporters and I hit the error:

```
$ pnpm -C test/core test basic.test.ts -- --reporter=json --reporter=default

> @vitest/test-core@ test /home/hiroshi/code/others/vitest/test/core
> vitest "basic.test.ts" "--reporter=json" "--reporter=default"

file:///home/hiroshi/code/others/vitest/packages/vitest/dist/cli.js:1208
      throw new Error(`Expected a single value for option "${command}"`);
            ^

Error: Expected a single value for option "--reporter <name>"
    at transform (file:///home/hiroshi/code/others/vitest/packages/vitest/dist/cli.js:1208:13)
    at setDotProp (file:///home/hiroshi/code/others/vitest/packages/vitest/dist/cli.js:258:22)
    at CAC.mri (file:///home/hiroshi/code/others/vitest/packages/vitest/dist/cli.js:642:9)
    at CAC.parse (file:///home/hiroshi/code/others/vitest/packages/vitest/dist/cli.js:559:27)
    at file:///home/hiroshi/code/others/vitest/packages/vitest/dist/cli.js:1342:13
    at ModuleJob.run (node:internal/modules/esm/module_job:218:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:329:24)
    at async main (file:///home/hiroshi/code/others/vitest/packages/vitest/dist/cli-wrapper.js:45:5)
```

Should we add `array: true` here to support this use case?

Comment (sheremet-va) [reply packages/vitest/src/node/cli/cli-config.ts:143]: Yes! And we can add a test for this 👍🏻

**Pull request merged**
